plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/XLanguageServerImpl.java

```diff
@@ -466,8 +466,13 @@ private void publishDiagnostics(URI uri, Iterable<? extends Issue> issues) {
 		initialized.thenAccept((initParams) -> {
 			PublishDiagnosticsParams publishDiagnosticsParams = new PublishDiagnosticsParams();
 			publishDiagnosticsParams.setUri(uriExtensions.toUriString(uri));
-			publishDiagnosticsParams.setDiagnostics(
-					workspaceManager.doRead(uri, (document, resource) -> toDiagnostics(issues, document)));
+			// this is not a premature optimization but a trick to handle issues of deleted resources
```

Comment:
actually, I also stumbled over this problem

Reply:
To me it's a strong indicator that the control flow is flawed.

Reply:
I think so too. I already changed this control flow on GH-1535 and created the ticket GH-1537 for later clean-ups.